src/map_io/map_buildingdata_packet.cc

```diff
@@ -819,17 +820,170 @@ void MapBuildingdataPacket::read_productionsite(ProductionSite& productionsite,
 				wq->read(fr, game, mol);
 
-				if (!game.descriptions().ware_exists(wq->get_index())) {
+				DescriptionIndex widx = wq->get_index();
+				if (!game.descriptions().ware_exists(widx)) {
+					++deleted_unknown_types;
+					deleted_unknown_amount += wq->get_filled();
 					delete wq;
+					inputs_changed = true;
 				} else {
+					// Savegame compatibility: check whether queue had size changed,
+					// or was removed altogether
+					auto it = std::find_if(
+					   curr_wares.begin(), curr_wares.end(), [widx](auto e) { return e.first == widx; });
+					if (it == curr_wares.end()) {
+						deleted_known.emplace_back(std::make_pair(widx, wq->get_filled()));
+						wq->set_filled(0u);
+						wq->cleanup();
+						delete wq;
+						inputs_changed = true;
+					} else {
+						const Quantity new_size = it->second;
+						const Quantity old_size = wq->get_max_size();
+						inputs_changed = inputs_changed || (new_size != old_size);
+						if (new_size > old_size) {
+							wq->set_max_size(new_size);
+							if (wq->get_max_fill() == old_size) {
+								wq->set_max_fill(new_size);
+							}
+							increased.emplace_back(game.descriptions().get_ware_descr(widx)->descname());
+						} else if (new_size < old_size) {
+							const Quantity old_filled = wq->get_filled();
+							Quantity lost = 0;
+							if (old_filled > new_size) {
+								lost = old_filled - new_size;
+								wq->set_filled(new_size);
+							}
+							wq->set_max_size(new_size);
+							decreased.emplace_back(std::make_pair(widx, lost));
+						}
+						productionsite.input_queues_.push_back(wq);
+					}
+				}
+			}
+			// Savegame compatibility: check for new queues that did not exist in older save file
+			for (WareAmount wa : curr_wares) {
+				DescriptionIndex widx = wa.first;
+				auto it = std::find_if(productionsite.input_queues_.begin(),
+				                       productionsite.input_queues_.end(),
+				                       [widx](auto e) { return e->get_index() == widx; });
+				if (it == productionsite.input_queues_.end()) {
+					WaresQueue* wq = new WaresQueue(productionsite, widx, wa.second);
 					productionsite.input_queues_.push_back(wq);
+					inputs_changed = true;
+					new_inputs.emplace_back(game.descriptions().get_ware_descr(widx)->descname());
 				}
 			}
 
+			// Report changes to the player
+			if (inputs_changed) {
+				const std::string title(_("Building’s inputs changed!"));
+
+				// Probably not worth adding graphic/text_layout as a dependency. It would
+				// require specifying the font styles too, but we already get that through
+				// Building::send_message().
+				// TODO(tothxa): The only problem is the hard-coded vspace gap.
+				static const std::string list_heading("<vspace gap=8><p>%s</p>");
+				static const std::string list_entry("<p> • %s</p>");
+
+				std::string body("<p>");
+				/** TRANSLATORS: %s is a building name */
+				body += format(_("The inputs of %s have changed. Please review the current "
```

Comment:
this won't work in all languages due to e.g. the article, sex , etc. of the buildings name,  
How about 
```suggestion
				body += format(_("The inputs of the building(s): "%s" have changed. Please review the current "
```